transport/transporters/s3/transporter/transporter.go

```diff
@@ -0,0 +1,351 @@
+/*
+  Copyright 2019 Nextdoor.com, Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+package transporter
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/Nextdoor/pg-bifrost.git/marshaller"
+	"github.com/Nextdoor/pg-bifrost.git/shutdown"
+	"github.com/Nextdoor/pg-bifrost.git/stats"
+	"github.com/Nextdoor/pg-bifrost.git/transport"
+	"github.com/Nextdoor/pg-bifrost.git/transport/batch"
+	"github.com/Nextdoor/pg-bifrost.git/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/cenkalti/backoff"
+	"github.com/cevaris/ordered_map"
+	"github.com/sirupsen/logrus"
+)
+
+// Generate mock for s3 client
+//go:generate mockgen -destination=mocks/mock_s3.go -package=mocks github.com/aws/aws-sdk-go/service/s3/s3iface S3API
+
+var (
+	TimeSource utils.TimeSource = utils.RealTime{}
+)
+
+// key_join is a helper to concatenate strings to form an S3 key
+func key_join(gzipped bool, strs ...string, ) string {
```

Comment:
Just a nit but this is the only part of the code that has gzip as optional.

Reply:
I was setting myself up for success in the future, but I can take it out as it's not strictly related to this PR.